server/eventlog/eventlog.go

```diff
@@ -23,41 +36,138 @@ func GetEventLogChunk(ctx context.Context, env environment.Env, req *elpb.GetEve
 		return &elpb.GetEventLogChunkResponse{
-			NextChunkId: fmt.Sprintf("%04x", 0),
+			Chunk: &elpb.GetEventLogChunkResponse_Chunk{
+				Buffer: make([]byte, 0),
+			},
 		}, nil
 	}
+	lastChunkIndex, err := chunkstore.ChunkIdAsUint16Index(lastChunkId)
+	if err != nil {
+		return nil, err
+	}
+
+	startIndex := lastChunkIndex
+	if req.ChunkId != "" {
+		var err error
+		if startIndex, err = chunkstore.ChunkIdAsUint16Index(req.ChunkId); err != nil {
+			return nil, err
+		}
+		if startIndex > lastChunkIndex {
+			if startIndex == math.MaxUint16 {
+				// The client requested the invalid id; this is an error.
+				return nil, status.ResourceExhaustedErrorf("Log index limit exceeded.")
+			}
+			// If out-of-bounds, return an empty chunk with PreviousChunkId set to the
+			// id of the last chunk.
+			rsp := &elpb.GetEventLogChunkResponse{
+				Chunk: &elpb.GetEventLogChunkResponse_Chunk{
+					Buffer: make([]byte, 0),
+				},
+				PreviousChunkId: chunkstore.ChunkIndexAsStringId(lastChunkIndex),
+			}
+			if invocationInProgress {
+				// If out-of-bounds and the invocation is in-progress, set NextChunkId
+				// to the id of the next chunk to be written.
+				rsp.NextChunkId = chunkstore.ChunkIndexAsStringId(lastChunkIndex + 1)
+			}
+			return rsp, nil
+		}
+	}
 
 	rsp := &elpb.GetEventLogChunkResponse{
 		Chunk: &elpb.GetEventLogChunkResponse_Chunk{
-			ChunkId: fmt.Sprintf("%04x", intChunkId),
+			ChunkId: chunkstore.ChunkIndexAsStringId(startIndex),
+			Buffer:  make([]byte, 0),
 		},
+		NextChunkId:     chunkstore.ChunkIndexAsStringId(startIndex + 1),
+		PreviousChunkId: chunkstore.ChunkIndexAsStringId(startIndex - 1),
 	}
-	if rsp.Chunk.Buffer, err = c.ReadChunk(ctx, GetEventLogPathFromInvocationId(req.InvocationId), intChunkId); err != nil {
-		return nil, err
+
+	boundary := lastChunkIndex
+	step := uint16(1)
+	if req.ChunkId == "" {
+		boundary = 0
+		step = math.MaxUint16 // decrements the value when added
 	}
-	if intChunkId > 0 {
-		rsp.PreviousChunkId = fmt.Sprintf("%04x", intChunkId-1)
+	lineCount := 0
+	// Fetch one chunk even if the minimum line count is 0
+	for chunkIndex := startIndex; chunkIndex != boundary+step; chunkIndex += step {
+		buffer, err := c.ReadChunk(ctx, eventLogPath, chunkIndex)
+		if err != nil {
+			return nil, err
+		}
+		scanner := bufio.NewScanner(bytes.NewReader(buffer))
+		for scanner.Scan() {
+			if scanner.Err() != nil {
+				return nil, err
+			}
+			lineCount++
+		}
+		if step == 1 {
+			rsp.Chunk.Buffer = append(rsp.Chunk.Buffer, buffer...)
+			rsp.NextChunkId = chunkstore.ChunkIndexAsStringId(chunkIndex + step)
+		} else {
+			rsp.Chunk.Buffer = append(buffer, rsp.Chunk.Buffer...)
+			rsp.PreviousChunkId = chunkstore.ChunkIndexAsStringId(chunkIndex + step)
+		}
+		if rsp.Chunk.ChunkId != chunkstore.ChunkIndexAsStringId(chunkIndex) {
+			// No longer fetching a single chunk, ChunkId cannot be meaningfully set.
+			rsp.Chunk.ChunkId = ""
+		}
```

Comment:
nit: Should the chunk ID field be removed from `Chunk`? The UI has no need for it.